rllib/agents/alpha_star/alpha_star.py

```diff
@@ -0,0 +1,592 @@
+                    NUM_AGENT_STEPS_TRAINED]
+
+        # For those policies that have been updated in this iteration
+        # (not all policies may have undergone an updated as we are
+        # requesting updates asynchronously):
+        # - Gather train infos.
+        # - Update weights to those remote rollout workers that contain
+        #   the respective policy.
+        with self._timers[SYNCH_WORKER_WEIGHTS_TIMER]:
+            train_infos = {}
+            policy_weights = {}
+            for pol_actor, policy_result in train_results.items():
+                if policy_result:
+                    pid = self.distributed_learners.get_policy_id(pol_actor)
+                    train_infos[pid] = policy_result
+                    policy_weights[pid] = pol_actor.get_weights.remote()
+
+            policy_weights_ref = ray.put(policy_weights)
+
+            for worker in self.workers.remote_workers():
+                worker.set_weights.remote(policy_weights_ref)
+
+        return train_infos
+
+    def build_league(self, result: ResultDict) -> None:
+        """Method containing league-building logic. Called after train step.
+
+        Args:
+            result: The most recent result dict with all necessary stats in
+                it (e.g. episode rewards) to perform league building
+                operations.
+        """
+
+        # If no evaluation results -> Use hist data gathered for training.
+        if "evaluation" in result:
+            hist_stats = result["evaluation"]["hist_stats"]
+        else:
+            hist_stats = result["hist_stats"]
+
+        trainable_policies = self.workers.local_worker().get_policies_to_train(
+        )
+        non_trainable_policies = \
+            set(self.workers.local_worker().policy_map.keys()) - \
+            trainable_policies
+
+        # Calculate current win-rates.
+        for policy_id, rew in hist_stats.items():
+            mo = re.match("^policy_(.+)_reward$", policy_id)
+            if mo is None:
+                continue
+            policy_id = mo.group(1)
+
+            # Calculate this policy's win rate.
+            won = 0
+            for r in rew:
+                if r > 0.0:  # win = 1.0; loss = -1.0
+                    won += 1
+            win_rate = won / len(rew)
+            # TODO: This should probably be a running average
+            #  (instead of hard-overriding it with the most recent data).
+            self.win_rates[policy_id] = win_rate
+
+            # Policy is a snapshot (frozen) -> Ignore.
+            if policy_id not in trainable_policies:
+                continue
+
+            print(
+                f"Iter={self.iteration} {policy_id}'s "
+                f"win-rate={win_rate} -> ",
+                end="")
+
+            # If win rate is good enough -> Snapshot current policy and decide,
+            # whether to freeze the new snapshot or not.
+            if win_rate >= self.config["win_rate_threshold_for_new_snapshot"]:
```

Comment:
I have a feeling these parameters should be per-agent, and not global.
although I am completely ok with starting with this for now and see.

Reply:
Probably. Then again, tons of things should probably be changed by the user depending on their league-building needs.